[Bug]: Storybook React renderer fails to compile with Typescript <5

### Describe the bug

Hi there! 

I just upgraded to Storybook 7 and hit the error:
 
> node_modules/@storybook/react/node_modules/type-fest/source/camel-case.d.ts:25:32 - error TS1005: '?' expected.
25  infer FirstWord extends string,
                                  ~
node_modules/@storybook/react/node_modules/type-fest/source/camel-case.d.ts:26:42 - error TS1005: '?' expected.
26  ...infer RemainingWords extends string[],
                                            ~
node_modules/@storybook/react/node_modules/type-fest/source/join.d.ts:59:36 - error TS1005: '?' expected.
59    infer First extends JoinableItem,
                                      ~
node_modules/@storybook/react/node_modules/type-fest/source/join.d.ts:60:49 - error TS1005: '?' expected.
60    ...infer Tail extends readonly JoinableItem[],
                                                   ~
node_modules/@storybook/react/node_modules/type-fest/source/join.d.ts:64:50 - error TS1005: '?' expected.
64     ...infer Head extends readonly JoinableItem[],
                                                    ~
node_modules/@storybook/react/node_modules/type-fest/source/join.d.ts:65:36 - error TS1005: '?' expected.
65     infer Last extends JoinableItem,
                                      ~
Found 6 errors.

Your package.json suggests you support any version of typescript
 
https://github.com/storybookjs/storybook/blob/5b34161518dd28b2451d2e075688a1c1e75928ef/code/renderers/react/package.json#L87

https://github.com/storybookjs/storybook/blob/5b34161518dd28b2451d2e075688a1c1e75928ef/code/renderers/react/package.json#L74


but [type-fest](https://www.npmjs.com/package/type-fest#:~:text=Requires%20TypeScript%20%3E%3D5.1) 
> Requires TypeScript >=5.1

I can't use Typescript 5 yet, could you pin Typescript to `^5` so people know older versions of typescript won't work, please? 

I have `skipLibCheck: true` so tbh I'm not sure why it's even bothering to check. 

### To Reproduce

I'm on `typescipt: ~4.5`, and am using React and vite. 
Default everything. 
Run `tsc` (i.e. `npm run compile` in the repro). 

https://stackblitz.com/edit/github-9ggjcp?file=tsconfig.json 

### System

```shell
Environment Info:

  System:
    OS: Windows 10 10.0.19045
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  Binaries:
    Node: 16.14.2 - C:\Program Files\nodejs\node.EXE
    npm: 8.19.3 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 114.0.5735.199
    Edge: Spartan (44.19041.1266.0), Chromium (114.0.1823.82)
  npmPackages:
    @storybook/addon-essentials: ^7.1.0 => 7.1.0
    @storybook/addon-interactions: ^7.1.0 => 7.1.0
    @storybook/addon-links: ^7.1.0 => 7.1.0
    @storybook/blocks: ^7.1.0 => 7.1.0
    @storybook/react: ^7.1.0 => 7.1.0
    @storybook/react-vite: ^7.1.0 => 7.1.0
    @storybook/testing-library: ^0.2.0 => 0.2.0
```


### Additional context

_No response_

--- a/code/lib/core-common/package.json
+++ b/code/lib/core-common/package.json
@@ -71,7 +71,7 @@
     "@types/picomatch": "^2.3.0",
     "mock-fs": "^5.2.0",
     "slash": "^5.0.0",
-    "type-fest": "~3.6.0",
+    "type-fest": "~2.19",
     "typescript": "~4.9.3"
   },
   "publishConfig": {

--- a/code/lib/core-common/package.json
+++ b/code/lib/core-common/package.json
@@ -71,7 +71,7 @@
     "@types/picomatch": "^2.3.0",
     "mock-fs": "^5.2.0",
     "slash": "^5.0.0",
-    "type-fest": "^3.11.0",
+    "type-fest": "~3.6.0",
     "typescript": "~4.9.3"
   },
   "publishConfig": {

--- a/code/package.json
+++ b/code/package.json
@@ -88,7 +88,7 @@
     "playwright": "1.36.0",
     "playwright-core": "1.36.0",
     "serialize-javascript": "^3.1.0",
-    "type-fest": "~3.6.0"
+    "type-fest": "~2.19"
   },
   "dependencies": {
     "@babel/core": "^7.22.9",

--- a/code/package.json
+++ b/code/package.json
@@ -88,7 +88,7 @@
     "playwright": "1.36.0",
     "playwright-core": "1.36.0",
     "serialize-javascript": "^3.1.0",
-    "type-fest": "^3.11.0"
+    "type-fest": "~3.6.0"
   },
   "dependencies": {
     "@babel/core": "^7.22.9",

--- a/code/renderers/react/package.json
+++ b/code/renderers/react/package.json
@@ -71,7 +71,7 @@
     "prop-types": "^15.7.2",
     "react-element-to-jsx-string": "^15.0.0",
     "ts-dedent": "^2.0.0",
-    "type-fest": "~3.6.0",
+    "type-fest": "~2.19",
     "util-deprecate": "^1.0.2"
   },
   "devDependencies": {

--- a/code/renderers/react/package.json
+++ b/code/renderers/react/package.json
@@ -71,7 +71,7 @@
     "prop-types": "^15.7.2",
     "react-element-to-jsx-string": "^15.0.0",
     "ts-dedent": "^2.0.0",
-    "type-fest": "^3.11.0",
+    "type-fest": "~3.6.0",
     "util-deprecate": "^1.0.2"
   },
   "devDependencies": {

--- a/code/renderers/svelte/package.json
+++ b/code/renderers/svelte/package.json
@@ -60,7 +60,7 @@
     "@storybook/preview-api": "7.2.0-alpha.0",
     "@storybook/types": "7.2.0-alpha.0",
     "sveltedoc-parser": "^4.2.1",
-    "type-fest": "^3.11.0"
+    "type-fest": "~3.6.0"
   },
   "devDependencies": {
     "expect-type": "^0.15.0",

--- a/code/renderers/svelte/package.json
+++ b/code/renderers/svelte/package.json
@@ -60,7 +60,7 @@
     "@storybook/preview-api": "7.2.0-alpha.0",
     "@storybook/types": "7.2.0-alpha.0",
     "sveltedoc-parser": "^4.2.1",
-    "type-fest": "~3.6.0"
+    "type-fest": "~2.19"
   },
   "devDependencies": {
     "expect-type": "^0.15.0",

--- a/code/renderers/vue/package.json
+++ b/code/renderers/vue/package.json
@@ -55,7 +55,7 @@
     "@storybook/preview-api": "7.2.0-alpha.0",
     "@storybook/types": "7.2.0-alpha.0",
     "ts-dedent": "^2.0.0",
-    "type-fest": "^3.11.0"
+    "type-fest": "~3.6.0"
   },
   "devDependencies": {
     "typescript": "~4.9.3",

--- a/code/renderers/vue/package.json
+++ b/code/renderers/vue/package.json
@@ -55,7 +55,7 @@
     "@storybook/preview-api": "7.2.0-alpha.0",
     "@storybook/types": "7.2.0-alpha.0",
     "ts-dedent": "^2.0.0",
-    "type-fest": "~3.6.0"
+    "type-fest": "~2.19"
   },
   "devDependencies": {
     "typescript": "~4.9.3",

--- a/code/renderers/vue3/package.json
+++ b/code/renderers/vue3/package.json
@@ -55,7 +55,7 @@
     "@storybook/types": "7.2.0-alpha.0",
     "lodash": "^4.17.21",
     "ts-dedent": "^2.0.0",
-    "type-fest": "~3.6.0",
+    "type-fest": "~2.19",
     "vue-component-type-helpers": "latest"
   },
   "devDependencies": {

--- a/code/renderers/vue3/package.json
+++ b/code/renderers/vue3/package.json
@@ -55,7 +55,7 @@
     "@storybook/types": "7.2.0-alpha.0",
     "lodash": "^4.17.21",
     "ts-dedent": "^2.0.0",
-    "type-fest": "^3.11.0",
+    "type-fest": "~3.6.0",
     "vue-component-type-helpers": "latest"
   },
   "devDependencies": {

--- a/scripts/package.json
+++ b/scripts/package.json
@@ -51,7 +51,7 @@
     "esbuild": "^0.18.0",
     "eslint": "^8.28.0",
     "serialize-javascript": "^3.1.0",
-    "type-fest": "~3.6.0"
+    "type-fest": "~2.19"
   },
   "dependencies": {
     "@actions/core": "^1.10.0",
@@ -179,7 +179,7 @@
     "ts-dedent": "^2.0.0",
     "ts-node": "^10.9.1",
     "tsup": "^6.7.0",
-    "type-fest": "~3.6.0",
+    "type-fest": "~2.19",
     "typescript": "5.0.4",
     "util": "^0.12.4",
     "uuid": "^9.0.0",

--- a/scripts/package.json
+++ b/scripts/package.json
@@ -51,7 +51,7 @@
     "esbuild": "^0.18.0",
     "eslint": "^8.28.0",
     "serialize-javascript": "^3.1.0",
-    "type-fest": "^3.11.0"
+    "type-fest": "~3.6.0"
   },
   "dependencies": {
     "@actions/core": "^1.10.0",
@@ -179,7 +179,7 @@
     "ts-dedent": "^2.0.0",
     "ts-node": "^10.9.1",
     "tsup": "^6.7.0",
-    "type-fest": "^3.11.0",
+    "type-fest": "~3.6.0",
     "typescript": "5.0.4",
     "util": "^0.12.4",
     "uuid": "^9.0.0",

